staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1467 @@
+		// be seen changing by this function are Status.IPv4 and Status.AddressVNI.
+		ca.queue.Add(parse.AttNSN(att))
+	}
+}
+
+func (ca *ConnectionAgent) onLocalAttRemoved(obj interface{}) {
+	peeledObj := parse.Peel(obj)
+	att := peeledObj.(*netv1a1.NetworkAttachment)
+	klog.V(5).Infof("Local NetworkAttachments cache: notified of removal of %#+v", att)
+	ca.queue.Add(parse.AttNSN(att))
+}
+
+func (ca *ConnectionAgent) waitForLocalAttsCacheSync(stopCh <-chan struct{}) error {
+	if !k8scache.WaitForCacheSync(stopCh, ca.localAttsInformer.HasSynced) {
+		return fmt.Errorf("caches failed to sync")
+	}
+	return nil
+}
+
+func (ca *ConnectionAgent) syncPreExistingIfcs() error {
+	if err := ca.syncPreExistingLocalIfcs(); err != nil {
+		return err
+	}
+
+	return ca.syncPreExistingRemoteIfcs()
+}
+
+func (ca *ConnectionAgent) syncPreExistingLocalIfcs() error {
+	allPreExistingLocalIfcs, err := ca.netFabric.ListLocalIfcs()
+	if err != nil {
+		return fmt.Errorf("failed initial local network interfaces list: %s", err.Error())
+	}
+
+	for _, aPreExistingLocalIfc := range allPreExistingLocalIfcs {
+		ifcVNIAndIP := localIfcVNIAndIP(&aPreExistingLocalIfc)
+		ifcOwnerAtts, err := ca.localAttsInformer.GetIndexer().ByIndex(attVNIAndIPIndexerName, ifcVNIAndIP)
+		if err != nil {
+			return fmt.Errorf("indexing local network interface with VNI/IP=%s failed: %s",
+				ifcVNIAndIP,
+				err.Error())
+		}
+
+		if len(ifcOwnerAtts) == 1 {
+			// A local attachment which should own the interface because their
+			// VNI and IP match was found.
+			ifcOwnerAtt := ifcOwnerAtts[0].(*netv1a1.NetworkAttachment)
+			nsn := parse.AttNSN(ifcOwnerAtt)
+			oldLocalState, oldIfcExists := ca.getLocalAttState(nsn)
+			ca.setLocalAttMainState(nsn, LocalAttachmentMainState{aPreExistingLocalIfc, ifcOwnerAtt.Spec.PostDeleteExec})
+			ca.setExecReport(nsn, ifcOwnerAtt.Status.PostCreateExecReport)
+			klog.V(3).Infof("Matched interface %#+v with local attachment %#+v", aPreExistingLocalIfc, ifcOwnerAtt)
+			if oldIfcExists {
+				aPreExistingLocalIfc = oldLocalState.LocalNetIfc
+			} else {
+				continue
+			}
+		}
+
+		// The interface must be deleted, e.g. because it could not be matched
+		// to an attachment, or because the attachment to which it has already
+		// been matched has changed and was matched to a different interface.
+		for i := 1; true; i++ {
+			err := ca.DeleteLocalIfc(aPreExistingLocalIfc)
+			if err == nil {
+				break
+			}
+			klog.V(3).Infof("Deletion of orphan local interface %#+v failed: %s. Attempt nbr. %d",
+				aPreExistingLocalIfc,
+				err.Error(),
+				i)
+			time.Sleep(netFabricRetryPeriod)
+		}
+		// Do not run the PostDeleteExec, it probably would not work
+		// because the PostCreateExec ran in a different container if
+		// at all.
+		klog.V(3).Infof("Deleted orphan local interface %#+v", aPreExistingLocalIfc)
+	}
+
+	return nil
+}
+
+func (ca *ConnectionAgent) syncPreExistingRemoteIfcs() error {
```

Comment:
We have a great deal of difficulty stemming from the use of the dynamic Indexers for reverse lookup.  There is a simpler approach.  Start the sync by using the Listers to enumerate all the NetworkAttachment objects and create a static reverse index, then do the rest of the sync work using that fixed reverse index.  Any changes that come in after the static reverse index was constructed will cause object references to go into the work queue and a worker will eventually catch up.

It will still be a little complicated because each Lister reveals an independent collection state, so there can be overlaps, but even so we can make fixed decisions with fixed data.  The overlaps will be on NetworkAttachment Namespace&Name, and I think it will be practical to resolve them first by creation timestamp (since a NA becomes interesting as a remote one only if it exists long enough to get implemented) and then by any tie-breaker we like.

Reply:
Regarding my remark about strictly ordering local sync before remote sync, indeed that was a bit too strong a statement.  What I really think is that the local NetworkAttachments have to be enumerated first in order to find all the relevant VNIs.

Note that the outlines you gave relied on unified data structures in the connection agent but not really on a unified API for the fabic.

You mentioned sharding for different data types but I think you meant sharding for one.  I think that we can expect that even when there is sharding to multiple etcd clusters for one data type, we can expect that if a non-cache-hitting LIST operation A strictly precedes non-cache-hitting LIST operation B then the results of B will all be at least as fresh as the results of A.

When an old match is rejected in favor of a newer match, I am not entirely happy deleting the old Linux network interface.  There is another approach that does not do this.  Add an outer loop.  The loop body starts with a given list of network interfaces to consider, and creates another list of network interfaces to reconsider.  When a match is rejected, the network interface involved is put on the list of network interfaces to reconsider.  At the end of the loop body, `interfacesToConsider = interfacesToReconsider`.  The loop continues as long as `len(interfacesToConsider) > 0`.

Yes, I think your correctly ordered outline would work.  And yes, I am concerned about the stacked up latencies.

I think it will be very rare for the initial sync to come across two matches for a given Linux network interface, so for the sake of simplicity we should probably satisfy ourselves with handling that in the simplest way that is correct.  So I think my net net is to use unified data structure in the connection agent and dynamic reverse lookups, delete a network interface if/when it is rejected from an old match.  This does not need a unified fabric API, and I still do not favor that.  As long as the connection agent uses unified data structures, the fabric API differences can be isolated to a few small places in the code (perhaps as method implementations).

Reply:
> You mentioned sharding for different data types but I think you meant sharding for one
Yes, sorry.

I see one problem with your solution based on a list of interfaces to reconsider: there's no guarantee that the loop terminates.

>  So I think my net net is to use unified data structure in the connection agent and dynamic reverse lookups, delete a network interface if/when it is rejected from an old match. This does not need a unified fabric API, and I still do not favor that. As long as the connection agent uses unified data structures, the fabric API differences can be isolated to a few small places in the code (perhaps as method implementations).

Good, I'm at work.